Add documentation for adaptive metrics and CRF score (#332)
scikit-learn-contrib/MAPIE: doc/theoretical_description_conformity_scores.rst

This score is by default symmetric (*see above for definition*).

2. The gamma score
==================

The gamma score (:class:`mapie.conformity_scores.GammaConformityScore`) adds a
notion of adaptivity with the normalization of the residuals by the predictions.

.. math:: \frac{|Y-\hat{\mu}(X)|}{\hat{\mu}(X)}

It computes adaptive intervals : intervals of different size on each example, with
the following formula  :

.. math:: [\hat{\mu}(X) * (1 - q(s)), \hat{\mu}(X) * (1 + q(s))]

Where :math:`q(s)` is the :math:`(1-\alpha)` quantile of the conformity scores.
(see :doc:`theoretical_description_regression` for more details).

This score is by default asymmetric (*see definition above*).

Compared to the absolute residual score, it allows to see regions with smaller intervals
than others which are interpreted as regions with more certainty than others.
It is important to note that, this conformity score is inversely proportional to the
order of magnitude of the predictions. Therefore, the uncertainty is proportional to
the order of magitude of the predictions, implying that this core should be used

> in use cases where we want greater uncertainty when the prediction is high.

Reviewer: Are these insights based on a paper? Or can we link to some example we have?

Author: This is the traduction of the score formula
I can link this paper : "Cordier, T., Blot, V., Lacombe, L., Morzadec, T., Capitaine, A. &amp; Brunel, N.. (2023). Flexible and Systematic Uncertainty Estimation with Conformal Prediction via the MAPIE library"

3. The conformal residual fitting score
=======================================

The conformal residual fitting score (:class:`mapie.conformity_scores.ConformalizedResidualFittingScore`)
(CRF) is slightly more complex than the previous scores.
The normalization of the residual is now done by the predictions of an additional model
:math:`\sigma` which learns to predict the base model residuals from :math:`X`.
:math:`\sigma` is trained on :math:`(X, |Y-\hat{\mu}(X)|)` and the formula of the score is :

.. math:: \frac{|Y-\hat{\mu}(X)|}{\hat{\sigma}(X)}

This score provides adaptive intervals : intervals of different sizes in each point
with the following formula :

.. math:: [\hat{\mu}(X) - q(s) * \hat{\sigma}(X), \hat{\mu}(X) + q(s) * \hat{\sigma}(X)]

Where :math:`q(s)` is the :math:`(1-\alpha)` quantile of the conformity scores.
(see :doc:`theoretical_description_regression` for more details).

This score is by default symmetric (*see definition above*). Unlike the scores above,
and due to the additionnal model required this score can only be used with split methods.

> Normalisation by the learned residuals from :math:`X` adds to the score a knowledge of

Reviewer: Is this from a paper?

Author: No but for now it is only implemented in the split setup

:math:`X` and its similarity to the other examples in the dataset.
Compared to the gamma score, the other adaptive score implemented in MAPIE,
it is not proportional to the uncertainty.

Key takeaways
=============

> - The absolute residual score is the basic conformity score and gives constant intervals.

Reviewer: Maybe it's worth mentioning that's in the basic one that is used throughout MAPIE as default?

Author: Done

.. title:: Theoretical Description : contents

.. _theoretical_description_conformity_scores:

=============================================

> Theoretical Description for Conformity Scores

Reviewer 2: ```suggestion
Theoretical Description for conformity scores
```

the following formula  :

.. math:: [\hat{\mu}(X) * (1 - q(s)), \hat{\mu}(X) * (1 + q(s))]

Where :math:`q(s)` is the :math:`(1-\alpha)` quantile of the conformity scores.
(see :doc:`theoretical_description_regression` for more details).

This score is by default asymmetric (*see definition above*).

Compared to the absolute residual score, it allows to see regions with smaller intervals
than others which are interpreted as regions with more certainty than others.
It is important to note that, this conformity score is inversely proportional to the
order of magnitude of the predictions. Therefore, the uncertainty is proportional to
the order of magitude of the predictions, implying that this score should be used
in use cases where we want greater uncertainty when the prediction is high.

3. The conformal residual fitting score
=======================================

The conformal residual fitting score [1] (:class:`mapie.conformity_scores.ConformalizedResidualFittingScore`)
(CRF) is slightly more complex than the previous scores.
The normalization of the residual is now done by the predictions of an additional model

> :math:`\sigma` which learns to predict the base model residuals from :math:`X`.
> :math:`\sigma` is trained on :math:`(X, |Y-\hat{\mu}(X)|)` and the formula of the score is :

Reviewer 2: ```suggestion
:math:`\hat\sigma` which learns to predict the base model residuals from :math:`X`.
:math:`\hat\sigma` is trained on :math:`(X, |Y-\hat{\mu}(X)|)` and the formula of the score is:
```

.. title:: Theoretical Description : contents

.. _theoretical_description_conformity_scores:

=========================================

> Theoretical Description Conformity Scores

Author: Done

Before describing the methods, let's briefly present the mathematical setting.
With conformal predictions we want to transform an heuristic notion of uncertainty
from a model into a rigorous one, and the first step to do it is to choose a conformal score.
The only requirement for the score function :math:`s(X, Y) \in \mathbb{R}` is
that larger scores should encode worse agreement between :math:`X` and :math:`Y`. [1]

There are two types of scores : the symmetric and asymmetric ones.
The symmetric property defines the way of computing the quantile of the conformity
scores when calculating the interval's bounds. If a score is symmetrical two
quantiles will be computed : one on the right side of the distribution
and the other on the left side.

1. The absolute residual score
==============================

The absolute residual score (:class:`mapie.conformity_scores.AbsoluteResidualScore`)
is the simplest and most commonly used conformal score, it translates the error
of the model : in regression it is called the residual.

.. math:: |Y-\hat{\mu}(X)|

With this score the intervals of predictions will be constant over the whole dataset.

> This score is symmetric (*see above for definition*).

Reviewer 2: ```suggestion
This score is by default symmetric (*see above for definition*).
```

Author: Done

2. The gamma score
==================

The gamma score (:class:`mapie.conformity_scores.GammaConformityScore`) adds a
notion of adaptivity with the normalization of the residuals by the predictions.

.. math:: \frac{|Y-\hat{\mu}(X)|}{\hat{\mu}(X)}

It computes adaptive intervals : intervals of different size on each example.

> This score is asymmetric (*see definition above*).

Reviewer 2: ```suggestion
This score is by default asymmetric (*see definition above*).
```

Author: Done

Compared to the absolute residual score, it allows to see regions with smaller intervals

> than others which are interpreted as regions with more certainty than others.

Reviewer 2: SUGGESTION: you can add more details about this adaptability. Uncertainties are proportional to the order of magnitude of the prediction (for example if $y=10$ then $\delta y = \pm 1$, if $y=100$ then $\delta y = \pm 10$). In other words, if my prediction is multiplied by 10, my uncertainties are also multiplied by 10.

Author: Done

Author: Done

3. The conformal residual fitting score
=======================================

The conformal residual fitting score (:class:`mapie.conformity_scores.ConformalizedResidualFittingScore`)
(CRF) is slightly more complex than the previous scores.
The normalization of the residual is now done by the predictions of an additional model
:math:`\sigma` which learns to predict the base model residuals from :math:`X`.

> :math:`\sigma` is trained on :math:`(X, |Y-\hat{\mu}(X)|)` and the formula of th score is :

Reviewer 2: ```suggestion
:math:`\sigma` is trained on :math:`(X, |Y-\hat{\mu}(X)|)` and the formula of the score is :
```

Author: Done

.. math:: \frac{|Y-\hat{\mu}(X)|}{\hat{\sigma}(X)}

This score provides adaptive intervals : intervals of different sizes in each point

> and is symmetric (*see definition above*). Unlike the scores above, and due to

Reviewer 2: ```suggestion
and is by default symmetric (*see definition above*). Unlike the scores above, and due to
```

the additionnal model required this score can only be used with split methods.

Normalisation by the learned residuals from :math:`X` adds to the score a knowledge of
:math:`X` and its similarity to the other examples in the dataset. In fact, using this
score results in even more adaptive intervals. Compared to the gamma score, the other adaptive
score implemented in MAPIE, it maintains relevant interval sizes over the entire dataset
even when there are outliers that could perturb the model. With gamma score, if
there are strong outliers, the intervals over the hole datset are sometimes too large

> to be useful.

Reviewer 2: I wouldn't be so categorical about the gamma and CRF method. When you talk about outliers here, you mean predictions that are not of the order of magnitude of the usual predictions, right? and I'm not sure I understand the last sentence.

Author: I agree, I wrote a new proposition.

the model is uncertain). Whereas, with the gamma score, the intervals would be large

> over the hole dataset.

Author: Done

Key takeaways
=============

- The absolute residual score is the basic conformity score and gives constant intervals.
- The gamma conformity score adds a notion of adaptivity by giving intervals of different sizes,

>   but it can give absurd results with strong outliers.

Reviewer 2: ```suggestion
  but it can give absurd results if the prediction model is too uncertain.
```

Author: I'm not sure to understand what you mean by "(if the uncertainty is of the order of magnitude of the predictions)"
